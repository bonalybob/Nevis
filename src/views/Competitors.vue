<template>
  <main>
    <back-arrow :to="`/events/${$route.params.id}`" />
    <h1 class="mx-10 mb-1 px-1">
      Entries
    </h1>
    <div class="mx-12 mb-3">
      <router-link
        :to="`/events/${$route.params.id}/competitors/create`"
        tag="button"
        class="button"
      >
        Create Entry
      </router-link>
      <button class="button" @click="refresh">Refresh</button>
    </div>
    <div
      v-if="allCompetitors && allCompetitors.length > 0"
      class="my-shadow mx-12 mb-5"
    >
      <h3 class="px-4 pt-2 text-blue select-none">
        Search Entries
      </h3>
      <text-input v-model="filterName" label="Name:" />
      <text-input v-model="filterSIID" label="SI Card:" />
      <dropdown-input
        v-model="filterCourse"
        :list="listOfCourseNames"
        label="Course:"
      />
    </div>
    <div
      v-if="competitors && competitors.length > 0"
      class="my-shadow mx-12 mb-3 p-2"
    >
      <table class="w-full font-body">
        <tr class="font-heading text-center hover:bg-blue-light">
          <th @click="changeSortBy('name')">
            Name
            <up-down-arrows
              :active="sortBy === 'name'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th @click="changeSortBy('siid')">
            SI Card
            <up-down-arrows
              :active="sortBy === 'siid'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th class="hidden md:table-cell" @click="changeSortBy('ageClass')">
            Age Class
            <up-down-arrows
              :active="sortBy === 'ageClass'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th @click="changeSortBy('course')">
            Course
            <up-down-arrows
              :active="sortBy === 'course'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
          <th class="hidden lg:table-cell" @click="changeSortBy('club')">
            Club
            <up-down-arrows
              :active="sortBy === 'club'"
              :ascending="sortDirection === 'ASC'"
            />
          </th>
        </tr>
        <tbody is="transition-group" name="fade">
          <router-link
            v-for="competitor of competitors"
            :key="competitor.id"
            :to="`/events/${$route.params.id}/competitors/${competitor.id}/edit`"
            tag="tr"
            class="text-center odd:bg-blue-lightest hover:bg-blue-light"
          >
            <td>{{ competitor.name }}</td>
            <td>{{ competitor.siid }}</td>
            <td class="hidden md:table-cell">{{ competitor.ageClass }}</td>
            <td>{{ competitor.courseName }}</td>
            <td class="hidden lg:table-cell">{{ competitor.club }}</td>
          </router-link>
        </tbody>
      </table>
    </div>
  </main>
</template>

<script>
import BackArrow from '@/components/BackArrow'
import TextInput from '@/components/TextInput'
import DropdownInput from '@/components/DropdownInput'
import UpDownArrows from '@/components/UpDownArrows'

export default {
  components: {
    'back-arrow': BackArrow,
    'text-input': TextInput,
    'dropdown-input': DropdownInput,
    'up-down-arrows': UpDownArrows,
  },

  data: function () {
    return {
      allCompetitors: [],
      competitors: [],
      listOfCourseNames: [],
      filterName: '',
      filterSIID: '',
      filterCourse: '',
      sortBy: 'id',
      sortDirection: 'DESC',
    }
  },

  watch: {
    filterName: function () {
      this.getCompetitors()
    },
    filterSIID: function () {
      this.getCompetitors()
    },
    filterCourse: function () {
      this.getCompetitors()
    },
  },

  mounted: function () {
    if (this.$database.connection === null || !this.$database.connected) {
      this.$router.push('/')
      this.$messages.addMessage('Problem Connecting To Database', 'error')
    } else this.refresh()
  },

  methods: {
    refresh: function () {
      this.getCourses()
      this.getCompetitors()
      this.getAllCompetitors()
    },
    changeSortBy: function (field) {
      if (this.sortBy === field && this.sortDirection === 'ASC')
        this.sortDirection = 'DESC'
      else if (this.sortBy === field) this.sortDirection = 'ASC'
      this.sortBy = field
      this.getCompetitors()
    },

    getCompetitors: function () {
      return this.$database
        .query(
          `
      SELECT competitors.*, courses.name AS courseName
      FROM competitors
      LEFT JOIN courses ON  courses.id=competitors.course
      WHERE courses.event=? AND competitors.event=?
        AND competitors.name LIKE '%${this.filterName}%'
        AND competitors.siid LIKE '%${this.filterSIID}%'
        AND courses.name LIKE '%${this.filterCourse}%'
      UNION
      SELECT competitors.*, '' AS courseName
      FROM competitors
      WHERE ISNULL(competitors.course)
        AND competitors.name LIKE '%${this.filterName}%'
        AND competitors.siid LIKE '%${this.filterSIID}%'
      ORDER BY ${this.sortBy} ${this.sortDirection}`,
          [this.$route.params.id, this.$route.params.id]
        )
        .then((result) => {
          this.competitors = result
        })
        .catch(() =>
          this.$messages.addMessage('Problem Fetching Entries', 'error')
        )
    },

    getAllCompetitors: function () {
      return this.$database
        .query(`SELECT * FROM competitors WHERE competitors.event=?`, [
          this.$route.params.id,
        ])
        .then((result) => {
          this.allCompetitors = result
        })
        .catch(() =>
          this.$messages.addMessage('Problem Fetching Entries', 'error')
        )
    },

    getCourses: function () {
      return this.$database
        .query('SELECT * FROM courses WHERE event=?', this.$route.params.id)
        .then((result) => {
          this.listOfCourseNames = result.map((course) => course.name)
          this.listOfCourseNames.unshift('')
        })
        .catch(() =>
          this.$messages.addMessage('Problem Fetching Courses', 'error')
        )
    },
  },
}
</script>
