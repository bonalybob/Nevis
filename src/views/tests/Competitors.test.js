import { mount, shallowMount } from '@vue/test-utils'
import Competitors from '@/views/Competitors'

test('Is a Vue Instance', () => {
  const wrapper = shallowMount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly - No Competitors', () => {
  const wrapper = shallowMount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Renders Correctly - With Competitors', () => {
  const wrapper = shallowMount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setData({
    competitors: [
      { id: 0, name: 'Bob', course: 1, siid: '154585' },
      { id: 77, name: 'Shaun', course: 2, siid: '12345' },
    ],
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Not Connected to the Database', () => {
  const wrapper = shallowMount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: false, query: jest.fn().mockResolvedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  expect(wrapper.vm.$router.push).toHaveBeenCalledTimes(1)
  expect(wrapper.vm.$router.push).toHaveBeenLastCalledWith('/')
})

test('Get Competitors - Success', async () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockResolvedValue(['hello']) },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.competitors).toEqual(['hello'])
})

test('Get Competitors - Error', async () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCompetitors()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Entries', 'error')
  expect(wrapper.vm.competitors.length).toBe(0)
})

test('Get Courses - Success', async () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: {
        connection: {},
        connected: true,
        query: jest.fn().mockResolvedValue([
          { id: 0, name: 'Long', length: 2.35, climb: 150, controls: '101,102,103' },
          { id: 77, name: 'Short', length: 1.75, climb: 15, controls: '100,108,103' },
        ]),
      },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.listOfCourseNames).toEqual(['', 'Long', 'Short'])
})

test('Get Courses - Error', async () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  await wrapper.vm.getCourses()
  expect(wrapper.vm.$messages.addMessage).toHaveBeenLastCalledWith('Problem Fetching Courses', 'error')
  expect(wrapper.vm.listOfCourseNames).toEqual([])
})

test('Watchers', () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setMethods({
    getCompetitors: jest.fn(),
  })
  wrapper.setData({ filterName: 'A' })
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(1)
  wrapper.setData({ filterName: 'B' })
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(2)
  wrapper.setData({ filterCourse: 'A' })
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(3)
  wrapper.setData({ filterSIID: 'A' })
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(4)
})

test('Change Sort By', () => {
  const wrapper = mount(Competitors, {
    stubs: ['router-link'],
    mocks: {
      $database: { connection: {}, connected: true, query: jest.fn().mockRejectedValue() },
      $route: { params: { id: 12 }, path: '' },
      $router: { push: jest.fn() },
      $messages: { addMessage: jest.fn() },
    },
  })
  wrapper.setMethods({
    getCompetitors: jest.fn(),
  })
  wrapper.vm.changeSortBy('name')
  expect(wrapper.vm.sortBy).toBe('name')
  expect(wrapper.vm.sortDirection).toBe('DESC')
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(1)
  wrapper.vm.changeSortBy('name')
  expect(wrapper.vm.sortBy).toBe('name')
  expect(wrapper.vm.sortDirection).toBe('ASC')
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(2)
  wrapper.vm.changeSortBy('siid')
  expect(wrapper.vm.sortBy).toBe('siid')
  expect(wrapper.vm.sortDirection).toBe('ASC')
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(3)
  wrapper.vm.changeSortBy('siid')
  expect(wrapper.vm.sortBy).toBe('siid')
  expect(wrapper.vm.sortDirection).toBe('DESC')
  expect(wrapper.vm.getCompetitors).toHaveBeenCalledTimes(4)
})
