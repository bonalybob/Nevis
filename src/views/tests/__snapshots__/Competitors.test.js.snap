// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly - No Competitors 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to="/events/12"
    />
    Entries
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/12/competitors/create"
    >
      
      Create Entry
    
    </router-link-stub>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <!---->
   
  <!---->
</main>
`;

exports[`Renders Correctly - With Competitors 1`] = `
<main>
  <h1
    class="mx-10 mb-1"
  >
    <back-arrow-stub
      to="/events/12"
    />
    Entries
  
  </h1>
   
  <div
    class="mx-12 mb-3"
  >
    <router-link-stub
      class="button"
      tag="button"
      to="/events/12/competitors/create"
    >
      
      Create Entry
    
    </router-link-stub>
     
    <button
      class="button"
    >
      Refresh
    </button>
  </div>
   
  <!---->
   
  <div
    class="shadow mx-12 mb-3 p-2"
  >
    <table
      class="w-full font-body"
    >
      <tr
        class="font-heading text-center hover:bg-blue-light"
      >
        <th>
          
          Name
          
          <up-down-arrows-stub />
        </th>
         
        <th>
          
          SI Card
          
          <up-down-arrows-stub />
        </th>
         
        <th
          class="hidden md:table-cell"
        >
          
          Age Class
          
          <up-down-arrows-stub />
        </th>
         
        <th>
          
          Course
          
          <up-down-arrows-stub />
        </th>
         
        <th
          class="hidden lg:table-cell"
        >
          
          Club
          
          <up-down-arrows-stub />
        </th>
      </tr>
       
      <tbody
        name="fade"
      >
        <router-link-stub
          class="text-center odd:bg-blue-lightest hover:bg-blue-light"
          tag="tr"
          to="/events/12/competitors/0/edit"
        >
          <td>
            Bob
          </td>
           
          <td>
            154585
          </td>
           
          <td
            class="hidden md:table-cell"
          >
            
          </td>
           
          <td>
            
          </td>
           
          <td
            class="hidden lg:table-cell"
          >
            
          </td>
        </router-link-stub>
        <router-link-stub
          class="text-center odd:bg-blue-lightest hover:bg-blue-light"
          tag="tr"
          to="/events/12/competitors/77/edit"
        >
          <td>
            Shaun
          </td>
           
          <td>
            12345
          </td>
           
          <td
            class="hidden md:table-cell"
          >
            
          </td>
           
          <td>
            
          </td>
           
          <td
            class="hidden lg:table-cell"
          >
            
          </td>
        </router-link-stub>
      </tbody>
    </table>
  </div>
</main>
`;
